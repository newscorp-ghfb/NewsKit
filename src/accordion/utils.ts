// JSDom doesn't support layout, and we can't manually set `scrollHeight` on elements
// until they are rendered. So accessing this prop directly makes testing difficult.
// Moving the logic to a standalone function that we can mock in tests solves this.
export const getRefScrollHeight = (el: HTMLDivElement) => el.scrollHeight;
