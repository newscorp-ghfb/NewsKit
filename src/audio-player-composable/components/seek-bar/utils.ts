export const formatTrackData = (
  trackColor: string,
  indicatorColor: string,
  bufferColor: string,
  timeArr: number[],
  buffered: TimeRanges | undefined,
) => {
  if (!buffered) {
    return {colors: [], values: []};
  }

  const time = timeArr[0];
  let bufferPositions = [];
  let bufferObjectArray = [];

  const {length} = buffered;
  for (let i = 0; i < length; i += 1) {
    bufferObjectArray.push({type: 'start', value: buffered.start(i)});
    bufferObjectArray.push({type: 'end', value: buffered.end(i)});
  }

  bufferObjectArray = bufferObjectArray.filter(val => val.value > time);
  if (bufferObjectArray[0] && bufferObjectArray[0].type === 'start') {
    bufferObjectArray.shift();
  }

  bufferPositions = bufferObjectArray.map(val => val.value);

  const colors = [
    indicatorColor,
    ...bufferPositions.map((x, i) => {
      if (i === 0) return bufferColor;
      return trackColor;
    }),
    trackColor,
  ];
  const values = [time, ...bufferPositions];
  return {colors, values};
};

export const seekBarAriaValueText = (seekTime: number[]) => {
  const time = new Date(seekTime[0] * 1000)
    .toISOString()
    .substr(11, 8)
    .split(':')
    .map(x => (+x < 10 ? (+x).toString() : x));

  const hour =
    time[0] !== '0' ? `${time[0]} ${time[0] === '1' ? 'hour' : 'hours'}` : '';
  const min =
    hour || time[1] !== '0'
      ? `${time[1]} ${time[1] === '1' ? 'minute' : 'minutes'}`
      : '';
  const sec = `${time[2]} ${time[2] === '1' ? 'second' : 'seconds'}`;

  return [hour, min, sec].filter(Boolean).join(' ');
};
