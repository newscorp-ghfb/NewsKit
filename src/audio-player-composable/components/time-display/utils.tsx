import {FormatFn} from './types';

export const calculateTime = (secs: number) => {
  const hours = Math.floor(secs / 3600);
  const minutes = Math.floor((secs % 3600) / 60);
  const seconds = Math.floor(secs % 60);
  const returnedSeconds = seconds < 10 ? `0${seconds}` : `${seconds}`;
  const returnedMinutes = minutes < 10 ? `0${minutes}` : `${minutes}`;
  const returnedHours = hours < 10 ? `0${hours}` : `${hours}`;
  if (hours > 0) {
    return `${returnedHours}:${returnedMinutes}:${returnedSeconds}`;
  }
  return `${returnedMinutes}:${returnedSeconds}`;
};
// change to formattedcurrentime
export const formatFunction: FormatFn = ({currentTime, duration}) => {
  const currentFormattedTime = calculateTime(
    (currentTime as unknown) as number,
  );
  const totalDuration = calculateTime((duration as unknown) as number);

  return `${currentFormattedTime}/${totalDuration} `;
};
