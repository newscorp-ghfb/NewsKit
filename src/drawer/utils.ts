const tabbableElements = [
  'a[href]',
  'area[href]',
  'input:not([disabled]):not([type="hidden"])',
  'select:not([disabled])',
  'textarea:not([disabled])',
  'button:not([disabled])',
  'iframe',
  'object',
  'embed',
  '[tabindex="0"]',
  '[contenteditable]',
  'audio[controls]',
  'video[controls]',
  'summary',
  '[tabindex^="0"]',
  '[tabindex^="1"]',
  '[tabindex^="2"]',
  '[tabindex^="3"]',
  '[tabindex^="4"]',
  '[tabindex^="5"]',
  '[tabindex^="6"]',
  '[tabindex^="7"]',
  '[tabindex^="8"]',
  '[tabindex^="9"]',
];

// Elements inside a container with aria-hidden set to true can't be focusable.
// This function aims to fix this issue in the drawer.
export const setDrawerElementFocusability = (
  open: boolean,
  ref: React.RefObject<HTMLDivElement>,
) => {
  /* istanbul ignore next */
  if (!ref || !ref.current) return;

  const dialog = ref.current;
  const focusableElements = dialog.querySelectorAll(tabbableElements.join());
  if (open) {
    focusableElements.forEach(element => {
      element.removeAttribute('tabindex');
    });
  }
  if (!open) {
    focusableElements.forEach(element => {
      element.setAttribute('tabindex', '-1');
    });
  }
};
