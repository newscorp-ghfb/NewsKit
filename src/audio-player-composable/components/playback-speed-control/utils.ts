import {BreakpointKeys, Theme} from '../../../theme';
import {deepMerge} from '../../../utils';
import {filterOutFalsyProperties} from '../../../utils/filter-object';
import {mergeBreakpointObject} from '../../../utils/merge-breakpoint-object';
import {AudioPlayerPlaybackSpeedControlOverridesProps} from './types';

export const iconButtonOverrides = (
  theme: Theme,
  overrides: AudioPlayerPlaybackSpeedControlOverridesProps,
) => ({
  ...deepMerge(
    mergeBreakpointObject(Object.keys(theme.breakpoints) as BreakpointKeys[]),
    theme.componentDefaults.audioPlayerPlaybackSpeedControl.iconButton,
    filterOutFalsyProperties(overrides.iconButton),
  ),
});

export const popoverOverrides = (
  theme: Theme,
  overrides: AudioPlayerPlaybackSpeedControlOverridesProps,
) => ({
  ...deepMerge(
    mergeBreakpointObject(Object.keys(theme.breakpoints) as BreakpointKeys[]),
    theme.componentDefaults.audioPlayerPlaybackSpeedControl.popover,
    filterOutFalsyProperties(overrides.popover),
  ),
});

export const selectionListOverrides = (
  theme: Theme,
  overrides: AudioPlayerPlaybackSpeedControlOverridesProps,
) => ({
  ...deepMerge(
    mergeBreakpointObject(Object.keys(theme.breakpoints) as BreakpointKeys[]),
    theme.componentDefaults.audioPlayerPlaybackSpeedControl.selectionList,
    filterOutFalsyProperties(overrides.selectionList),
  ),
});

export const selectionListOptionOverrides = (
  theme: Theme,
  overrides: AudioPlayerPlaybackSpeedControlOverridesProps,
) => ({
  ...deepMerge(
    mergeBreakpointObject(Object.keys(theme.breakpoints) as BreakpointKeys[]),
    theme.componentDefaults.audioPlayerPlaybackSpeedControl.selectionListOption,
    filterOutFalsyProperties(overrides.selectionListOptions),
  ),
});

export const modalOverrides = (
  theme: Theme,
  overrides: AudioPlayerPlaybackSpeedControlOverridesProps,
) => ({
  ...deepMerge(
    mergeBreakpointObject(Object.keys(theme.breakpoints) as BreakpointKeys[]),
    theme.componentDefaults.audioPlayerPlaybackSpeedControl.modal,
    filterOutFalsyProperties(overrides.modal),
  ),
});
