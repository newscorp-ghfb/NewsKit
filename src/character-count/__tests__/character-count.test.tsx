import React, {ReactElement, RefObject} from 'react';
import {act} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import {CharacterCount} from '..';
import {generateString, renderWithImplementation} from '../../test/test-utils';
import {useRefWithReRender} from '../../utils/use-ref-with-rerender';
import {TextArea} from '../../text-area';

const InputWithCharacterCount = ({
  children,
}: {
  children: (ref: RefObject<any>) => ReactElement;
}) => {
  const ref = useRefWithReRender<HTMLTextAreaElement>(null);
  return children(ref);
};

describe('CharacterCount', () => {
  const MAX_LENGTH = 100;
  const MIN_LENGTH = 20;
  const MSG = 'some text';

  test('displays no message if no max or min specified', () => {
    const {getByTestId, asFragment} = renderWithImplementation(
      InputWithCharacterCount,
      {
        children: ref => (
          <>
            <TextArea ref={ref} data-testid="text-area" />
            <CharacterCount inputRef={ref} data-testid="character-count" />
          </>
        ),
      },
    );
    const characterCount = getByTestId('character-count');
    expect(characterCount.textContent).toEqual(``);
    expect(asFragment()).toMatchSnapshot();
  });

  test('applies overrides', () => {
    const {asFragment} = renderWithImplementation(InputWithCharacterCount, {
      children: ref => (
        <>
          <TextArea ref={ref} data-testid="text-area" minLength={MIN_LENGTH} />
          <CharacterCount
            inputRef={ref}
            data-testid="character-count"
            overrides={{
              marginBlock: 'space040',
              paddingInline: 'space060',
              minHeight: '80px',
            }}
          />
        </>
      ),
    });
    expect(asFragment()).toMatchSnapshot();
  });

  test('message updates if both min and max specified', async () => {
    const {getByTestId} = renderWithImplementation(InputWithCharacterCount, {
      children: ref => (
        <>
          <TextArea
            minLength={MIN_LENGTH}
            maxLength={MAX_LENGTH}
            ref={ref}
            data-testid="text-area"
          />
          <CharacterCount inputRef={ref} data-testid="character-count" />
        </>
      ),
    });
    const characterCount = getByTestId('character-count');
    expect(characterCount.textContent).toEqual(
      `Please enter a minimum of ${MIN_LENGTH} characters`,
    );
    const textArea = getByTestId('text-area');
    await act(async () => {
      await userEvent.type(textArea, generateString(MIN_LENGTH));
    });
    expect(characterCount.textContent).toEqual(
      `You have ${MAX_LENGTH - MIN_LENGTH} characters remaining`,
    );
  });

  test('accepts custom format function', () => {
    const {getByTestId} = renderWithImplementation(InputWithCharacterCount, {
      children: ref => (
        <>
          <TextArea
            minLength={MIN_LENGTH}
            maxLength={MAX_LENGTH}
            ref={ref}
            data-testid="text-area"
          />
          <CharacterCount
            inputRef={ref}
            data-testid="character-count"
            format={({currentLength, maxLength, minLength}) =>
              `This input has current length ${currentLength}, min ${minLength} and max ${maxLength}`
            }
          />
        </>
      ),
    });
    const characterCount = getByTestId('character-count');
    expect(characterCount.textContent).toEqual(
      `This input has current length 0, min ${MIN_LENGTH} and max ${MAX_LENGTH}`,
    );
  });

  describe('with maxLength', () => {
    test('displays starting number of characters remaining', async () => {
      const {getByTestId, asFragment} = renderWithImplementation(
        InputWithCharacterCount,
        {
          children: ref => (
            <>
              <TextArea
                ref={ref}
                data-testid="text-area"
                maxLength={MAX_LENGTH}
              />
              <CharacterCount inputRef={ref} data-testid="character-count" />
            </>
          ),
        },
      );
      const characterCount = getByTestId('character-count');
      expect(characterCount.textContent).toEqual(
        `You have ${MAX_LENGTH} characters remaining`,
      );
      expect(asFragment()).toMatchSnapshot();
    });

    test('displays starting number of characters remaining with default value', async () => {
      const {getByTestId} = renderWithImplementation(InputWithCharacterCount, {
        children: ref => (
          <>
            <TextArea
              ref={ref}
              defaultValue={MSG}
              data-testid="text-area"
              maxLength={MAX_LENGTH}
            />
            <CharacterCount inputRef={ref} data-testid="character-count" />
          </>
        ),
      });
      const characterCount = getByTestId('character-count');
      expect(characterCount.textContent).toEqual(
        `You have ${MAX_LENGTH - MSG.length} characters remaining`,
      );
    });

    test('updates number of characters remaining', async () => {
      const {getByTestId} = renderWithImplementation(InputWithCharacterCount, {
        children: ref => (
          <>
            <TextArea
              ref={ref}
              data-testid="text-area"
              maxLength={MAX_LENGTH}
            />
            <CharacterCount inputRef={ref} data-testid="character-count" />
          </>
        ),
      });
      const characterCount = getByTestId('character-count');
      const textArea = getByTestId('text-area');
      await act(async () => {
        await userEvent.type(textArea, MSG);
      });
      expect(characterCount.textContent).toEqual(
        `You have ${MAX_LENGTH - MSG.length} characters remaining`,
      );
    });

    test('uses singular when only one character remains', async () => {
      const {getByTestId} = renderWithImplementation(InputWithCharacterCount, {
        children: ref => (
          <>
            <TextArea
              ref={ref}
              data-testid="text-area"
              maxLength={MAX_LENGTH}
            />
            <CharacterCount inputRef={ref} data-testid="character-count" />
          </>
        ),
      });
      const characterCount = getByTestId('character-count');
      const textArea = getByTestId('text-area');
      await act(async () => {
        await userEvent.type(textArea, generateString(MAX_LENGTH - 1));
      });
      expect(characterCount.textContent).toEqual(
        `You have 1 character remaining`,
      );
    });
  });

  describe('with minLength', () => {
    test('displays starting number of characters required', async () => {
      const {getByTestId, asFragment} = renderWithImplementation(
        InputWithCharacterCount,
        {
          children: ref => (
            <>
              <TextArea
                ref={ref}
                data-testid="text-area"
                minLength={MIN_LENGTH}
              />
              <CharacterCount inputRef={ref} data-testid="character-count" />
            </>
          ),
        },
      );
      const characterCount = getByTestId('character-count');
      expect(characterCount.textContent).toEqual(
        `Please enter a minimum of ${MIN_LENGTH} characters`,
      );
      expect(asFragment()).toMatchSnapshot();
    });
    test('displays singular when only one character required', async () => {
      const {getByTestId} = renderWithImplementation(InputWithCharacterCount, {
        children: ref => (
          <>
            <TextArea ref={ref} data-testid="text-area" minLength={1} />
            <CharacterCount inputRef={ref} data-testid="character-count" />
          </>
        ),
      });
      const characterCount = getByTestId('character-count');
      expect(characterCount.textContent).toEqual(
        `Please enter a minimum of 1 character`,
      );
    });

    test('displays starting number of characters required with default value', async () => {
      const {getByTestId} = renderWithImplementation(InputWithCharacterCount, {
        children: ref => (
          <>
            <TextArea
              ref={ref}
              defaultValue={MSG}
              data-testid="text-area"
              minLength={MIN_LENGTH}
            />
            <CharacterCount inputRef={ref} data-testid="character-count" />
          </>
        ),
      });
      const characterCount = getByTestId('character-count');
      expect(characterCount.textContent).toEqual(
        `Please enter ${MIN_LENGTH - MSG.length} characters`,
      );
    });

    test('updates number of characters remaining', async () => {
      const {getByTestId} = renderWithImplementation(InputWithCharacterCount, {
        children: ref => (
          <>
            <TextArea
              ref={ref}
              data-testid="text-area"
              minLength={MIN_LENGTH}
            />
            <CharacterCount inputRef={ref} data-testid="character-count" />
          </>
        ),
      });
      const characterCount = getByTestId('character-count');
      const textArea = getByTestId('text-area');
      await act(async () => {
        await userEvent.type(textArea, MSG);
      });
      expect(characterCount.textContent).toEqual(
        `Please enter ${MIN_LENGTH - MSG.length} characters`,
      );
    });

    test('does not show when minimum is met', async () => {
      const {getByTestId} = renderWithImplementation(InputWithCharacterCount, {
        children: ref => (
          <>
            <TextArea
              ref={ref}
              data-testid="text-area"
              minLength={MIN_LENGTH}
            />
            <CharacterCount inputRef={ref} data-testid="character-count" />
          </>
        ),
      });
      const characterCount = getByTestId('character-count');
      const textArea = getByTestId('text-area');
      await act(async () => {
        await userEvent.type(textArea, generateString(MIN_LENGTH));
      });
      expect(characterCount.textContent).toEqual('');
    });

    test('uses singular when only one more character is required', async () => {
      const {getByTestId} = renderWithImplementation(InputWithCharacterCount, {
        children: ref => (
          <>
            <TextArea
              ref={ref}
              data-testid="text-area"
              minLength={MIN_LENGTH}
            />
            <CharacterCount inputRef={ref} data-testid="character-count" />
          </>
        ),
      });
      const characterCount = getByTestId('character-count');
      const textArea = getByTestId('text-area');
      await act(async () => {
        await userEvent.type(textArea, generateString(MIN_LENGTH - 1));
      });
      expect(characterCount.textContent).toEqual(`Please enter 1 character`);
    });
  });
});
