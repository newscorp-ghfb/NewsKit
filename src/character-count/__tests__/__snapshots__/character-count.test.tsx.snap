// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterCount applies overrides 1`] = `
<DocumentFragment>
  .emotion-0 {
  box-sizing: border-box;
  cursor: default;
  resize: vertical;
  width: 100%;
  height: 120px;
  min-height: 48px;
  background-color: transparent;
  border-style: solid;
  border-color: #919191;
  border-width: 1px;
  border-radius: 8px;
  color: #3B3B3B;
  text-overflow: ellipsis;
  font-family: "DM Sans",sans-serif;
  font-size: 14px;
  line-height: 21px;
  font-weight: 400;
  letter-spacing: 0;
  padding: 0.5px 0px;
  margin-block-end: 8px;
  padding-inline: 12px;
  padding-block: 12px;
}

.emotion-0 svg {
  fill: #3B3B3B;
}

.emotion-0:focus:not(:disabled) {
  outline-color: #3768FB;
  outline-style: solid;
  outline-width: 2px;
  border-color: transparent;
  background-color: #F1F1F1;
}

@media not all and (min-resolution: 0.001dpcm) {
  @supports (-webkit-appearance: none) and (stroke-color: transparent) {
    .emotion-0:focus:not(:disabled) {
      outline-style: auto;
    }
  }
}

.emotion-0:hover:not(:disabled) {
  background-color: #F1F1F1;
}

.emotion-0:valid:focus:not(:disabled) {
  outline-color: #3768FB;
  outline-style: solid;
  outline-width: 2px;
  border-color: transparent;
  background-color: #F1F1F1;
}

@media not all and (min-resolution: 0.001dpcm) {
  @supports (-webkit-appearance: none) and (stroke-color: transparent) {
    .emotion-0:valid:focus:not(:disabled) {
      outline-style: auto;
    }
  }
}

.emotion-0:disabled {
  border-color: #E2E2E2;
  color: #ABABAB;
}

.emotion-0:disabled svg {
  fill: #ABABAB;
}

.emotion-0:invalid:focus:not(:disabled) {
  outline-color: #3768FB;
  outline-style: solid;
  outline-width: 2px;
  border-color: transparent;
  background-color: #F1F1F1;
}

@media not all and (min-resolution: 0.001dpcm) {
  @supports (-webkit-appearance: none) and (stroke-color: transparent) {
    .emotion-0:invalid:focus:not(:disabled) {
      outline-style: auto;
    }
  }
}

.emotion-0::before {
  content: '';
  margin-bottom: -0.391em;
  display: block;
}

.emotion-0::after {
  content: '';
  margin-top: -0.409em;
  display: block;
}

.emotion-0::-webkit-input-placeholder {
  color: #6A6A6A;
}

.emotion-0::-moz-placeholder {
  color: #6A6A6A;
}

.emotion-0:-ms-input-placeholder {
  color: #6A6A6A;
}

.emotion-0::placeholder {
  color: #6A6A6A;
}

<textarea
    class="emotion-0"
    data-testid="text-area"
    minlength="20"
  />
  .emotion-0 {
  margin: 0;
  margin-block: 16px;
  padding-inline: 32px;
  color: #6A6A6A;
  font-family: "DM Sans",sans-serif;
  font-size: 14px;
  line-height: 21px;
  font-weight: 400;
  letter-spacing: 0;
  padding: 0.5px 0px;
  min-height: 80px;
  margin-block-end: 8px;
  margin-block: 16px;
  padding-inline: 32px;
}

.emotion-0:disabled {
  color: #ABABAB;
}

.emotion-0::before {
  content: '';
  margin-bottom: -0.391em;
  display: block;
}

.emotion-0::after {
  content: '';
  margin-top: -0.409em;
  display: block;
}

<p
    class="emotion-0"
    data-testid="character-count"
  >
    Please enter a minimum of 20 characters
  </p>
</DocumentFragment>
`;

exports[`CharacterCount displays no message if no max or min specified 1`] = `
<DocumentFragment>
  .emotion-0 {
  box-sizing: border-box;
  cursor: default;
  resize: vertical;
  width: 100%;
  height: 120px;
  min-height: 48px;
  background-color: transparent;
  border-style: solid;
  border-color: #919191;
  border-width: 1px;
  border-radius: 8px;
  color: #3B3B3B;
  text-overflow: ellipsis;
  font-family: "DM Sans",sans-serif;
  font-size: 14px;
  line-height: 21px;
  font-weight: 400;
  letter-spacing: 0;
  padding: 0.5px 0px;
  margin-block-end: 8px;
  padding-inline: 12px;
  padding-block: 12px;
}

.emotion-0 svg {
  fill: #3B3B3B;
}

.emotion-0:focus:not(:disabled) {
  outline-color: #3768FB;
  outline-style: solid;
  outline-width: 2px;
  border-color: transparent;
  background-color: #F1F1F1;
}

@media not all and (min-resolution: 0.001dpcm) {
  @supports (-webkit-appearance: none) and (stroke-color: transparent) {
    .emotion-0:focus:not(:disabled) {
      outline-style: auto;
    }
  }
}

.emotion-0:hover:not(:disabled) {
  background-color: #F1F1F1;
}

.emotion-0:valid:focus:not(:disabled) {
  outline-color: #3768FB;
  outline-style: solid;
  outline-width: 2px;
  border-color: transparent;
  background-color: #F1F1F1;
}

@media not all and (min-resolution: 0.001dpcm) {
  @supports (-webkit-appearance: none) and (stroke-color: transparent) {
    .emotion-0:valid:focus:not(:disabled) {
      outline-style: auto;
    }
  }
}

.emotion-0:disabled {
  border-color: #E2E2E2;
  color: #ABABAB;
}

.emotion-0:disabled svg {
  fill: #ABABAB;
}

.emotion-0:invalid:focus:not(:disabled) {
  outline-color: #3768FB;
  outline-style: solid;
  outline-width: 2px;
  border-color: transparent;
  background-color: #F1F1F1;
}

@media not all and (min-resolution: 0.001dpcm) {
  @supports (-webkit-appearance: none) and (stroke-color: transparent) {
    .emotion-0:invalid:focus:not(:disabled) {
      outline-style: auto;
    }
  }
}

.emotion-0::before {
  content: '';
  margin-bottom: -0.391em;
  display: block;
}

.emotion-0::after {
  content: '';
  margin-top: -0.409em;
  display: block;
}

.emotion-0::-webkit-input-placeholder {
  color: #6A6A6A;
}

.emotion-0::-moz-placeholder {
  color: #6A6A6A;
}

.emotion-0:-ms-input-placeholder {
  color: #6A6A6A;
}

.emotion-0::placeholder {
  color: #6A6A6A;
}

<textarea
    class="emotion-0"
    data-testid="text-area"
  />
  .emotion-0 {
  margin: 0;
  color: #6A6A6A;
  font-family: "DM Sans",sans-serif;
  font-size: 14px;
  line-height: 21px;
  font-weight: 400;
  letter-spacing: 0;
  padding: 0.5px 0px;
  min-height: 12px;
  margin-block-end: 8px;
  padding-inline: 8px;
}

.emotion-0:disabled {
  color: #ABABAB;
}

.emotion-0::before {
  content: '';
  margin-bottom: -0.391em;
  display: block;
}

.emotion-0::after {
  content: '';
  margin-top: -0.409em;
  display: block;
}

<p
    class="emotion-0"
    data-testid="character-count"
  />
</DocumentFragment>
`;

exports[`CharacterCount displays state passed 1`] = `
<DocumentFragment>
  .emotion-0 {
  box-sizing: border-box;
  cursor: default;
  resize: vertical;
  width: 100%;
  height: 120px;
  min-height: 48px;
  background-color: transparent;
  border-style: solid;
  border-color: #919191;
  border-width: 1px;
  border-radius: 8px;
  color: #3B3B3B;
  text-overflow: ellipsis;
  font-family: "DM Sans",sans-serif;
  font-size: 14px;
  line-height: 21px;
  font-weight: 400;
  letter-spacing: 0;
  padding: 0.5px 0px;
  margin-block-end: 8px;
  padding-inline: 12px;
  padding-block: 12px;
}

.emotion-0 svg {
  fill: #3B3B3B;
}

.emotion-0:focus:not(:disabled) {
  outline-color: #3768FB;
  outline-style: solid;
  outline-width: 2px;
  border-color: transparent;
  background-color: #F1F1F1;
}

@media not all and (min-resolution: 0.001dpcm) {
  @supports (-webkit-appearance: none) and (stroke-color: transparent) {
    .emotion-0:focus:not(:disabled) {
      outline-style: auto;
    }
  }
}

.emotion-0:hover:not(:disabled) {
  background-color: #F1F1F1;
}

.emotion-0:valid:focus:not(:disabled) {
  outline-color: #3768FB;
  outline-style: solid;
  outline-width: 2px;
  border-color: transparent;
  background-color: #F1F1F1;
}

@media not all and (min-resolution: 0.001dpcm) {
  @supports (-webkit-appearance: none) and (stroke-color: transparent) {
    .emotion-0:valid:focus:not(:disabled) {
      outline-style: auto;
    }
  }
}

.emotion-0:disabled {
  border-color: #E2E2E2;
  color: #ABABAB;
}

.emotion-0:disabled svg {
  fill: #ABABAB;
}

.emotion-0:invalid:focus:not(:disabled) {
  outline-color: #3768FB;
  outline-style: solid;
  outline-width: 2px;
  border-color: transparent;
  background-color: #F1F1F1;
}

@media not all and (min-resolution: 0.001dpcm) {
  @supports (-webkit-appearance: none) and (stroke-color: transparent) {
    .emotion-0:invalid:focus:not(:disabled) {
      outline-style: auto;
    }
  }
}

.emotion-0::before {
  content: '';
  margin-bottom: -0.391em;
  display: block;
}

.emotion-0::after {
  content: '';
  margin-top: -0.409em;
  display: block;
}

.emotion-0::-webkit-input-placeholder {
  color: #6A6A6A;
}

.emotion-0::-moz-placeholder {
  color: #6A6A6A;
}

.emotion-0:-ms-input-placeholder {
  color: #6A6A6A;
}

.emotion-0::placeholder {
  color: #6A6A6A;
}

<textarea
    class="emotion-0"
    data-testid="text-area"
    maxlength="100"
  />
  .emotion-0 {
  margin: 0;
  color: #D60000;
  font-family: "DM Sans",sans-serif;
  font-size: 14px;
  line-height: 21px;
  font-weight: 400;
  letter-spacing: 0;
  padding: 0.5px 0px;
  min-height: 12px;
  margin-block-end: 8px;
  padding-inline: 8px;
}

.emotion-0::before {
  content: '';
  margin-bottom: -0.391em;
  display: block;
}

.emotion-0::after {
  content: '';
  margin-top: -0.409em;
  display: block;
}

<p
    class="emotion-0"
    data-testid="character-count"
  >
    You have 100 characters remaining
  </p>
</DocumentFragment>
`;

exports[`CharacterCount with maxLength displays starting number of characters remaining 1`] = `
<DocumentFragment>
  .emotion-0 {
  box-sizing: border-box;
  cursor: default;
  resize: vertical;
  width: 100%;
  height: 120px;
  min-height: 48px;
  background-color: transparent;
  border-style: solid;
  border-color: #919191;
  border-width: 1px;
  border-radius: 8px;
  color: #3B3B3B;
  text-overflow: ellipsis;
  font-family: "DM Sans",sans-serif;
  font-size: 14px;
  line-height: 21px;
  font-weight: 400;
  letter-spacing: 0;
  padding: 0.5px 0px;
  margin-block-end: 8px;
  padding-inline: 12px;
  padding-block: 12px;
}

.emotion-0 svg {
  fill: #3B3B3B;
}

.emotion-0:focus:not(:disabled) {
  outline-color: #3768FB;
  outline-style: solid;
  outline-width: 2px;
  border-color: transparent;
  background-color: #F1F1F1;
}

@media not all and (min-resolution: 0.001dpcm) {
  @supports (-webkit-appearance: none) and (stroke-color: transparent) {
    .emotion-0:focus:not(:disabled) {
      outline-style: auto;
    }
  }
}

.emotion-0:hover:not(:disabled) {
  background-color: #F1F1F1;
}

.emotion-0:valid:focus:not(:disabled) {
  outline-color: #3768FB;
  outline-style: solid;
  outline-width: 2px;
  border-color: transparent;
  background-color: #F1F1F1;
}

@media not all and (min-resolution: 0.001dpcm) {
  @supports (-webkit-appearance: none) and (stroke-color: transparent) {
    .emotion-0:valid:focus:not(:disabled) {
      outline-style: auto;
    }
  }
}

.emotion-0:disabled {
  border-color: #E2E2E2;
  color: #ABABAB;
}

.emotion-0:disabled svg {
  fill: #ABABAB;
}

.emotion-0:invalid:focus:not(:disabled) {
  outline-color: #3768FB;
  outline-style: solid;
  outline-width: 2px;
  border-color: transparent;
  background-color: #F1F1F1;
}

@media not all and (min-resolution: 0.001dpcm) {
  @supports (-webkit-appearance: none) and (stroke-color: transparent) {
    .emotion-0:invalid:focus:not(:disabled) {
      outline-style: auto;
    }
  }
}

.emotion-0::before {
  content: '';
  margin-bottom: -0.391em;
  display: block;
}

.emotion-0::after {
  content: '';
  margin-top: -0.409em;
  display: block;
}

.emotion-0::-webkit-input-placeholder {
  color: #6A6A6A;
}

.emotion-0::-moz-placeholder {
  color: #6A6A6A;
}

.emotion-0:-ms-input-placeholder {
  color: #6A6A6A;
}

.emotion-0::placeholder {
  color: #6A6A6A;
}

<textarea
    class="emotion-0"
    data-testid="text-area"
    maxlength="100"
  />
  .emotion-0 {
  margin: 0;
  color: #6A6A6A;
  font-family: "DM Sans",sans-serif;
  font-size: 14px;
  line-height: 21px;
  font-weight: 400;
  letter-spacing: 0;
  padding: 0.5px 0px;
  min-height: 12px;
  margin-block-end: 8px;
  padding-inline: 8px;
}

.emotion-0:disabled {
  color: #ABABAB;
}

.emotion-0::before {
  content: '';
  margin-bottom: -0.391em;
  display: block;
}

.emotion-0::after {
  content: '';
  margin-top: -0.409em;
  display: block;
}

<p
    class="emotion-0"
    data-testid="character-count"
  >
    You have 100 characters remaining
  </p>
</DocumentFragment>
`;

exports[`CharacterCount with minLength displays starting number of characters required 1`] = `
<DocumentFragment>
  .emotion-0 {
  box-sizing: border-box;
  cursor: default;
  resize: vertical;
  width: 100%;
  height: 120px;
  min-height: 48px;
  background-color: transparent;
  border-style: solid;
  border-color: #919191;
  border-width: 1px;
  border-radius: 8px;
  color: #3B3B3B;
  text-overflow: ellipsis;
  font-family: "DM Sans",sans-serif;
  font-size: 14px;
  line-height: 21px;
  font-weight: 400;
  letter-spacing: 0;
  padding: 0.5px 0px;
  margin-block-end: 8px;
  padding-inline: 12px;
  padding-block: 12px;
}

.emotion-0 svg {
  fill: #3B3B3B;
}

.emotion-0:focus:not(:disabled) {
  outline-color: #3768FB;
  outline-style: solid;
  outline-width: 2px;
  border-color: transparent;
  background-color: #F1F1F1;
}

@media not all and (min-resolution: 0.001dpcm) {
  @supports (-webkit-appearance: none) and (stroke-color: transparent) {
    .emotion-0:focus:not(:disabled) {
      outline-style: auto;
    }
  }
}

.emotion-0:hover:not(:disabled) {
  background-color: #F1F1F1;
}

.emotion-0:valid:focus:not(:disabled) {
  outline-color: #3768FB;
  outline-style: solid;
  outline-width: 2px;
  border-color: transparent;
  background-color: #F1F1F1;
}

@media not all and (min-resolution: 0.001dpcm) {
  @supports (-webkit-appearance: none) and (stroke-color: transparent) {
    .emotion-0:valid:focus:not(:disabled) {
      outline-style: auto;
    }
  }
}

.emotion-0:disabled {
  border-color: #E2E2E2;
  color: #ABABAB;
}

.emotion-0:disabled svg {
  fill: #ABABAB;
}

.emotion-0:invalid:focus:not(:disabled) {
  outline-color: #3768FB;
  outline-style: solid;
  outline-width: 2px;
  border-color: transparent;
  background-color: #F1F1F1;
}

@media not all and (min-resolution: 0.001dpcm) {
  @supports (-webkit-appearance: none) and (stroke-color: transparent) {
    .emotion-0:invalid:focus:not(:disabled) {
      outline-style: auto;
    }
  }
}

.emotion-0::before {
  content: '';
  margin-bottom: -0.391em;
  display: block;
}

.emotion-0::after {
  content: '';
  margin-top: -0.409em;
  display: block;
}

.emotion-0::-webkit-input-placeholder {
  color: #6A6A6A;
}

.emotion-0::-moz-placeholder {
  color: #6A6A6A;
}

.emotion-0:-ms-input-placeholder {
  color: #6A6A6A;
}

.emotion-0::placeholder {
  color: #6A6A6A;
}

<textarea
    class="emotion-0"
    data-testid="text-area"
    minlength="20"
  />
  .emotion-0 {
  margin: 0;
  color: #6A6A6A;
  font-family: "DM Sans",sans-serif;
  font-size: 14px;
  line-height: 21px;
  font-weight: 400;
  letter-spacing: 0;
  padding: 0.5px 0px;
  min-height: 12px;
  margin-block-end: 8px;
  padding-inline: 8px;
}

.emotion-0:disabled {
  color: #ABABAB;
}

.emotion-0::before {
  content: '';
  margin-bottom: -0.391em;
  display: block;
}

.emotion-0::after {
  content: '';
  margin-top: -0.409em;
  display: block;
}

<p
    class="emotion-0"
    data-testid="character-count"
  >
    Please enter a minimum of 20 characters
  </p>
</DocumentFragment>
`;
